### PR TITLE
abci: Port `EventAttribute` field type change to `main`

### LINKED ISSUE #6403: Events get mangled when non-utf8 value in Value
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):

v0.34.10

**ABCI app** (name for built-in, URL for self-written if it's publicly available):

Cosmos SDK v0.42.4 (and master)

**Environment**:
- **OS** (e.g. from /etc/os-release):
- **Install tools**:
- **Others**:

**What happened**:

A custom IBC app dared to actually return binary data in the acknowledgement, which was converted to a string `string(rawBytes)` for the cosmos sdk event, then back to `[]byte` for abci, and then into the tendermint rpc event system. Somewhere along the way it got mangled and a few bytes changed - which caused the protobuf data to not decode.

Relevant issue and fix (hex-encode binary before passing to event system) here: https://github.com/cosmos/ibc-go/pull/144

**What you expected to happen**:

If the abci.Event takes key and value as `[]byte` it should support them, no?

**Have you tried the latest version**: yes

**How to reproduce it** (as minimally and precisely as possible):

See https://github.com/cosmos/ibc-go/pull/144

**Suggestion**

Just make the `KVPair` used for returning Events over abci take `key string`, `value string`. There is no real reason for this and lots of code unsafely converts between strings and bytes. (I doubt any indexer handles raw binary well).

Once it used to be a nicely typed `string` -> `string | int` mapping

```proto
message KVPair {
  string key = 1;
  enum Type {
    STRING = 0;
    INT = 1;
  }
  Type value_type = 2;
  string value_string = 3;
  int64 value_int = 4;
}
```

But it was later converted to

```proto
message KVPair {
  bytes key = 1;
  bytes value = 2;
}
```

in https://github.com/tendermint/abci/pull/167. Although the arguments don't really convince me (as no processing code knows what to do with `[]byte` besides try to cast it):

> -> Protobuf3 strings are supposed to be only utf8 or 7-bit ansii but for something called "KVPair", we don't want to be restricted to unicode or ansii characters.
> -> Perhaps using []byte makes us more vigilant in validating the input before accepting it, than be lulled into complacency w/ a string.

This actually broke all searches in the event system, as it didn't know how to handle anything anymore. You can see a desire to revert this change here: https://github.com/tendermint/tendermint/issues/1369 (particularly comment https://github.com/tendermint/tendermint/issues/1369#issuecomment-377169038) and it seemed to have been decided to use `strings` again: https://github.com/tendermint/tendermint/issues/1369#issuecomment-383884334

Although the only fix was that the event system *expects all tags to be strings* https://github.com/tendermint/tendermint/pull/1498. It uses `map[string]string` to store the data that is abci-encoded as `[]byte`. WHY?

Please update the abci interface to use strings. This inspired refactoring is causing issues more than 3 years later.

### PR DESCRIPTION
**Pull request by thanethomson**
Closes #6403.

This PR is a port of #6408 to `main` for release in v0.37.

---

#### PR checklist

- [x] Tests written/updated, or no tests needed
- [x] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [x] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

File: state/txindex/kv/kv.go [hunk 1]
```diff
@@ -314,29 +315,30 @@ func (txi *TxIndex) match(
...
 
+	EQ_LOOP:
 		for ; it.Valid(); it.Next() {
 			tmpHashes[string(it.Value())] = it.Value()
 
 			// Potentially exit early.
 			select {
 			case <-ctx.Done():
-				break
+				break EQ_LOOP
```
Comment (sergio-mena) [inline state/txindex/kv/kv.go:333]: Are we sure that, if the context passed as parameter to the function is "done" we just want to break out of the loop? (but still execute the one at the end: REMOVE_LOOP)

Comment (thanethomson) [reply state/txindex/kv/kv.go:333]: Given the comment above the `select` statement I assumed that this is what was originally meant. Without specifying the `for` loop label, the existing code just breaks out of the `select` statement, which means we may as well just ignore the context altogether.

Comment (thanethomson) [reply state/txindex/kv/kv.go:333]: The only alternative I can think of is to return from the function with an error.

What would trigger the termination of the context outside of the tests?

Comment (sergio-mena) [reply state/txindex/kv/kv.go:333]: Yes, I got the intention of you changes, and they are definitely improving the logic.
It's just that,  if the context is "done", don't we just want to return immediately? (if we're unsure, let's just leave it as you have it)

**Review by sergio-mena — APPROVED** [review]
Managing contexts in this part of the code, anyway, needs much more work, I don't think it should block this PR

File: state/txindex/kv/kv.go (hunk 1 above)
Comment (thanethomson) [reply state/txindex/kv/kv.go:333]: Yeah I think this is something we need to clean up. I doubt it's catastrophic if we don't, but it is a bit messy right now.

**Pull request merged**